src/mscorlib/shared/System/Threading/Tasks/ValueTask.cs

```diff
@@ -283,30 +256,73 @@ public ValueTaskSourceTask(IValueTaskSource source, short token)
         public bool IsCompleted
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => _obj == null || (ObjectIsTask ? UnsafeGetTask().IsCompleted : UnsafeGetValueTaskSource().GetStatus(_token) != ValueTaskSourceStatus.Pending);
+            get
+            {
+                object obj = _obj;
```

Comment:
> object obj = _obj; [](start = 15, length = 19)

would be nice to add some comment why we are holding the value on local variable and using that local variable instead of _obj

Reply:
Ok.  I'll add a comment.  It's because we want to avoid the case where the _obj is replaced by something else during the operation, e.g. by someone writing to the same field such that a subsequent _obj read actually returns a different object.

Reply:
yes, I know the reason as I know what you are trying to fix here :-) I just wanted to ensure anyone else in the future reading the code understand why we are doing that.

Reply:
Is it guaranteed without a volatile load that uses of the local 'obj' would not load from the original memory location? The C++ compiler does this sometimes and it has caused issues (for instance, [this](https://github.com/dotnet/coreclr/pull/16089/commits/71251146b5515b016630e172fc70b2beabf2cc65#diff-c514f5790766cc71de26cd6d7f322127R432)). What about in .NET Native? Is there some rule in C# that prevents that?

Reply:
Technically I think it's possible on some architectures, but that would lead to a huge amount of code  potentially failing, as every place we do:
```C#
Foo f = _foo;
if (f != null) f.Whatever();
```
would be subject to a NullReferenceException.  That pattern happens all over the place, especially with C#'s `?.` operator, which is specifically billed as being usable in such multithreaded scenarios.

I'm hesitant to introduce a Volatile.Read here, though I can if it's deemed necessary.  @vancem / @CarolEidt, I think you've commented on such cases in the past.  Any thoughts?

Reply:
I have in the past suggested that replacing a single heap reference with multiple heap references is an optimization that we want to NOT do on a cost-benefit point of view.   In addition to the case that Stephen mentions above, there is also the issue with += operator that C# EMITS FOR YOU which has a similar issue (as well as most cost that call events, because they check for null and then call the method).  

So as a practical matter, we have already crossed this bridge (we have many places in our code base where 'optimizing' by 'refetching' from the heap will break things).     Also the value of such optimization is very low (especially if your architecture has many registers (which is everything but 32 bit X86)).   

I think it is OK to skip the Volatile.Read.   I just can't imagine us trying to do such a optimization (and we are probably simply going to turn off the optimization (if we have not already) in and native compiler we have.